packages/grid/x-data-grid-pro/src/utils/tree/sortRowTree.ts

```diff
@@ -68,24 +151,22 @@ export const sortRowTree = (params: SortRowTreeParams) => {
 
   sortGroup(rowTree[GRID_ROOT_GROUP_ID] as GridGroupNode);
 
-  // Flatten the sorted lists to have children just after their parent
-  const insertRowListIntoSortedRows = (startIndex: number, rowList: GridRowId[]) => {
-    sortedRows = [...sortedRows.slice(0, startIndex), ...rowList, ...sortedRows.slice(startIndex)];
+  const rootList = List.from<GridRowId>(sortedGroupedByParentRows.get(GRID_ROOT_GROUP_ID)!);
+
+  const fillList = (nodes: Node<GridRowId>[]) => {
+    nodes.forEach((node) => {
+      const children = sortedGroupedByParentRows.get(node.data);
 
-    let treeSize = 0;
-    rowList.forEach((rowId) => {
-      treeSize += 1;
-      const children = sortedGroupedByParentRows.get(rowId);
       if (children?.length) {
-        const subTreeSize = insertRowListIntoSortedRows(startIndex + treeSize, children);
-        treeSize += subTreeSize;
+        const sublist = List.from(children);
+        const subnodes = sublist.nodes();
+        node.insertAfter(sublist);
+        fillList(subnodes);
```

Comment:
I think that if I was iterating the list in the reverse order, then it would be possible to fill it without storing the list into a temporary `nodes` array.

Reply:
Not 100% sure, but if we use `List.forEach()` instead of `List.nodes().forEach()`, we  don't even need recursion here:

```ts
const rootList = List.from<GridRowId>(sortedGroupedByParentRows.get(GRID_ROOT_GROUP_ID)!);

rootList.forEach((node) => {
  const children = sortedGroupedByParentRows.get(node.data);

  if (children?.length) {
    const sublist = List.from(children);
    node.insertAfter(sublist);
  }
});
```

Reply:
Done

Reply:
No recursion single-linked list, 100,000 rows:

![image](https://github.com/mui/mui-x/assets/1423607/3240552f-5f92-4d0f-8cb2-62e024114eb0)

Reply:
There was indeed no need for recursion nor inversion. I was just very tired when I wrote this.